instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/net/PeerServiceResolver.java

```diff
@@ -0,0 +1,121 @@
+
+  PeerServiceResolver(Map<String, String> peerServiceMapping) {
+    peerServiceMapping.forEach(
+        (key, serviceName) -> {
+          try {
+            String protocol = null;
+            URI uri;
+            if (key.contains("://")) {
+              uri = new URI(key);
+              protocol = uri.getScheme();
+            } else {
+              uri = new URI("http://" + key);
+            }
+            String host = uri.getHost();
+            Integer port = uri.getPort() == -1 ? null : uri.getPort();
+            Map<ServiceMatcher, String> matchers =
+                MAPPING.computeIfAbsent(host, x -> new ConcurrentHashMap<>());
+            matchers.putIfAbsent(new ServiceMatcher(protocol, port, uri.getPath()), serviceName);
+          } catch (URISyntaxException use) {
+            // TODO wire logging
+          }
+        });
+  }
+
+  boolean isEmpty() {
+    return MAPPING.isEmpty();
+  }
+
+  String resolveService(String protocol, String host, Integer port, String path) {
+    Map<ServiceMatcher, String> matchers = MAPPING.get(host);
+    if (matchers == null) {
+      return null;
+    }
+    return matchers.entrySet().stream()
+        .filter(entry -> entry.getKey().matches(protocol, port, path))
+        .max(Comparator.comparingInt(entry -> entry.getKey().getPathLength()))
+        .map(Map.Entry::getValue)
+        .orElse(null);
+  }
+
+  private static class ServiceMatcher {
+
+    private final String protocol;
+    private final Integer port;
+    private final String path;
+
+    public ServiceMatcher(String protocol, Integer port, String path) {
+      this.protocol = protocol;
+      this.port = port;
+      this.path = path;
+    }
+
+    public int getPathLength() {
+      if (this.path == null) {
+        return 0;
+      }
+      return this.path.length();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof ServiceMatcher)) {
+        return false;
+      }
+      ServiceMatcher that = (ServiceMatcher) o;
+      return Objects.equals(protocol, that.protocol)
+          && Objects.equals(port, that.port)
+          && Objects.equals(path, that.path);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(protocol, port, path);
+    }
+
+    public boolean matches(String protocol, Integer port, String path) {
+      if (this.protocol != null) {
+        if (!this.protocol.equals(protocol)) {
+          return false;
+        }
+      }
+      if (this.port != null) {
+        if (!this.port.equals(port)) {
+          return false;
+        }
+      }
+      if (this.path != null) {
+        if (path == null) {
+          return false;
+        }
+        if (!path.startsWith(this.path)) {
+          return false;
+        }
+        if (port != null) {
+          return port.equals(this.port);
+        }
```

Comment:
Why is a `port` condition here? Isn't this part about the `path`?

Reply:
Yes it is. When the matcher has a path, the tested port need to match the matcher's port. Otherwise with the mapping :
localhost/mypath=myservice 

localhost:4859/mypath and localhost:8081/mypath would both be resolved as myservice.

Reply:
Yeah, but doesn't the
```java
        if (port != null) {
          return port.equals(this.port);
        }
```
condition above already handle that?

Reply:
The checker can have a null port.
The first port checking requires the two ports to match if the checker has a defined port.

When it comes to path checking.
Once the path are matching, then the ports must also be equals.

Reply:
If you have the mapping `(1.2.3.4, null, /abc) -> def` but in reality you have `(1.2.3.4, 8080, /abc) -> def` and `(1.2.3.4, 8081, /abc) -> ghi`, I believe that it's better to return `null` so the mapping has to be fixed to have the port provided instead of having a misleading mapping.